pkg/replica/hash.go

```diff
@@ -0,0 +1,340 @@
+	ModTime      string `json:"modTime"`
+	LastHashedAt string `json:"lastHashedAt"`
+}
+
+func NewSnapshotHashTask(ctx context.Context, cancel context.CancelFunc, snapshotName string, rehash bool) *SnapshotHashTask {
+	return &SnapshotHashTask{
+		Ctx:          ctx,
+		CancelFunc:   cancel,
+		SnapshotName: snapshotName,
+		Rehash:       rehash,
+
+		SnapshotHashStatus: SnapshotHashStatus{
+			State: ProgressStateInProgress,
+		},
+	}
+}
+
+func (t *SnapshotHashTask) LockFile() (fileLock *flock.Flock, err error) {
+	defer func() {
+		if err != nil && fileLock != nil && fileLock.Path() != "" {
+			if err := os.RemoveAll(fileLock.Path()); err != nil {
+				logrus.Warnf("failed to remove lock file %v since %v", fileLock.Path(), err)
+			}
+		}
+	}()
+
+	hostName := os.Getenv("HOSTNAME")
+	if hostName == "" {
+		return nil, fmt.Errorf("env HOSTNAME is missing")
+	}
+
+	err = os.MkdirAll(fileLockDirectory, 0755)
+	if err != nil {
+		return nil, err
+	}
+
+	fileLock = flock.New(filepath.Join(fileLockDirectory, hashLockFileName))
+
+	// Blocking lock
+	err = fileLock.Lock()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to fetch the file lock for hashing snapshot %v", t.SnapshotName)
+	}
+
+	return fileLock, nil
+}
+
+func (t *SnapshotHashTask) UnlockFile(fileLock *flock.Flock) {
+	fileLock.Unlock()
+}
+
+func (t *SnapshotHashTask) Execute() {
+	var err error
+	var checksum string
+
+	defer func() {
+		t.StatusLock.Lock()
+		defer t.StatusLock.Unlock()
+		t.Checksum = checksum
+		if err != nil {
+			logrus.Errorf("failed to hash snapshot %v since %v", t.SnapshotName, err)
+			t.State = ProgressStateError
+			t.Error = err.Error()
+		} else {
+			t.State = ProgressStateComplete
+		}
+	}()
+
+	// Each node can have only snapshot hashing task at the same time.
+	// When the snapshot hashing task is initiated, the task will try to fetch the
+	// file lock (${fileLockDirectory}/hash), and other tasks need to wait for the lock.
+	// The file is unlocked after the task is completed.
+	fileLock, err := t.LockFile()
+	if err != nil {
+		return
+	}
+	defer t.UnlockFile(fileLock)
+
+	modTime, err := GetSnapshotModTime(t.SnapshotName)
+	if err != nil {
+		return
+	}
+
+	requireRehash := true
+	if !t.Rehash {
+		requireRehash, checksum = t.isRehashRequired(modTime)
+		if !requireRehash {
+			return
+		}
+	}
+
+	logrus.Infof("Starting hashing snapshot %v", t.SnapshotName)
+
+	lastHashedAt := time.Now().UTC().Format(time.RFC3339)
+	checksum, err = hashSnapshot(t.Ctx, t.SnapshotName)
+	if err != nil {
+		return
+	}
+
+	if t.isSnapshotSilentlyCorrupted(checksum) {
+		t.StatusLock.Lock()
+		t.SilentlyCorrupted = true
+		t.StatusLock.Unlock()
+		return
+	}
+
+	logrus.Infof("Snapshot %v checksum %v", t.SnapshotName, checksum)
+
+	err = SetSnapshotHashInfoToXattr(t.SnapshotName, checksum, modTime, lastHashedAt)
+	if err != nil {
+		return
+	}
+
+	err = t.isModTimeRemain(modTime)
+	if err != nil {
+		err = DeleteSnapshotHashInfoFromXattr(t.SnapshotName)
+	}
```

Comment:
NIT: execute this before `SetSnapshotHashInfoToXattr`

Reply:
`SetSnapshotHashInfoToXattr` should override the xattr, so the `DeleteSnapshotHashInfoFromXattr` might be not necessary.
`DeleteSnapshotHashInfoFromXattr` here is just to delete the useless information.